core/test/config/utils.hpp

```diff
@@ -0,0 +1,90 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2023, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_TEST_CONFIG_UTILS_HPP_
+#define GKO_CORE_TEST_CONFIG_UTILS_HPP_
+
+
+#include <ostream>
+
+
+#include <ginkgo/core/config/data.hpp>
+#include <ginkgo/core/config/property_tree.hpp>
+
+namespace gko {
+namespace config {
+
+std::ostream& operator<<(std::ostream& stream, const data& d)
+{
+    if (holds_alternative<std::string>(d)) {
+        stream << '"' << get<std::string>(d) << '"';
+    } else if (holds_alternative<long long int>(d)) {
+        stream << get<long long int>(d);
+    } else if (holds_alternative<double>(d)) {
+        stream << get<double>(d);
+    } else if (holds_alternative<bool>(d)) {
+        stream << (get<bool>(d) ? "true" : "false");
+    } else if (holds_alternative<monostate>(d)) {
+        stream << "<empty>";
+    }
+    return stream;
+}
```

Comment:
both `data` and `pnode` could use a `visit` function implementing the Visitor pattern, which is commonly used together with variants. that would make accessing a generic object a bit more, well, generic :smile: An example can be seen [here](https://dev.to/tmr232/that-overloaded-trick-overloading-lambdas-in-c17)